tests/integration/tokenomics/relay_mining_integration_test.go

```diff
@@ -1,143 +1,214 @@
 )
 
-type RelayMiningIntegrationTestSuite struct {
-	suites.ParamsSuite
-}
+func TestComputeNewDifficultyHash_RewardsReflectWorkCompleted(t *testing.T) {
+	// Update the target number of relays to a value that suits the test.
+	// A too high number would make the difficulty stay at BaseRelayDifficultyHash
+	initialTargetRelays := servicekeeper.TargetNumRelays
+	servicekeeper.TargetNumRelays = 1000
+	t.Cleanup(func() {
+		// Reset the target number of relays to its initial value.
+		servicekeeper.TargetNumRelays = initialTargetRelays
+	})
+
+	// Prepare the test service.
+	service := sharedtypes.Service{
+		Id:                   "svc1",
+		Name:                 "svcName1",
+		ComputeUnitsPerRelay: 1,
+		OwnerAddress:         sample.AccAddress(),
+	}
 
-func (s *RelayMiningIntegrationTestSuite) SetupTest() {
-	// Construct a fresh integration app for each test.
-	s.NewApp(s.T())
-	s.SetupTestAuthzAccounts(s.T())
-	s.SetupTestAuthzGrants(s.T())
-}
+	// Prepare the test application.
+	appAddress := sample.AccAddress()
+	appStake := sdk.NewInt64Coin(volatile.DenomuPOKT, 1000)
+	application := apptypes.Application{
+		Address: appAddress,
+		Stake:   &appStake,
+		ServiceConfigs: []*sharedtypes.ApplicationServiceConfig{
+			{ServiceId: service.Id},
+		},
+	}
+
+	// Prepare the test supplier.
+	supplierAddress := sample.AccAddress()
+	supplierStake := sdk.NewInt64Coin(volatile.DenomuPOKT, 1000)
+	supplier := sharedtypes.Supplier{
+		OperatorAddress: supplierAddress,
+		OwnerAddress:    supplierAddress,
+		Stake:           &supplierStake,
+		Services: []*sharedtypes.SupplierServiceConfig{
+			{
+				ServiceId: service.Id,
+				RevShare: []*sharedtypes.ServiceRevenueShare{
+					{
+						Address:            supplierAddress,
+						RevSharePercentage: 100,
+					},
+				},
+			},
+		},
+	}
+
+	keepers, ctx := testutils.NewTokenomicsModuleKeepers(t, nil,
+		testutils.WithService(service),
+		testutils.WithApplication(application),
+		testutils.WithSupplier(supplier),
+	)
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	sdkCtx = sdkCtx.WithBlockHeight(1)
+
+	// Set the CUTTM to 1 to simplify the math
+	sharedParams := keepers.SharedKeeper.GetParams(sdkCtx)
+	sharedParams.ComputeUnitsToTokensMultiplier = uint64(1)
+	err := keepers.SharedKeeper.SetParams(sdkCtx, sharedParams)
+	require.NoError(t, err)
+
+	// Set the global proof params so we never need a proof (for simplicity of this test)
+	err = keepers.ProofKeeper.SetParams(sdkCtx, prooftypes.Params{
+		ProofRequestProbability:   0,                                                                            // we never need a proof randomly
+		ProofRequirementThreshold: &sdk.Coin{Denom: volatile.DenomuPOKT, Amount: sdkmath.NewInt(math.MaxInt64)}, // a VERY high threshold
+	})
+	require.NoError(t, err)
+
+	// Update the relay mining difficulty so there's always a difficulty to retrieve
+	// for the test service.
+	_, err = keepers.ServiceKeeper.UpdateRelayMiningDifficulty(sdkCtx, map[string]uint64{service.Id: 1})
```

Comment:
I'm noticing that this map is iterated over directly in `UpdateRelayMiningDifficulty()` where the loop body makes getter and setter calls to the keeper. Similar to the discussion we had about `SettlePendingClaims()`/`ProcessesTokenLogicModules()`, we should prefer to avoid using maps like this. 

Perhaps the general rule of thumb should be to accompany any map intended to be iterated with a slice of ordered keys. Perhaps we can formalize this as a generic type or something.

Would you mind adding a TODO to the `UpdatesRelayMiningDifficulty()` definition regarding this?

Reply:
Added a sort keys function that we should generalize.